File: android/app/src/main/java/ru/bgtu_voenmeh/zapara/data/profiles/ProfileDescriptor.kt
package ru.bgtu_voenmeh.zapara.data.profiles

data class ProfileDescriptor(
    val isGuest: Boolean,
    val userId: String?,
    val serverKey: String?,
    val databaseName: String
) {
    companion object {
        const val GUEST_DB = "zapara.db"

        fun guest() = ProfileDescriptor(true, null, null, GUEST_DB)

        fun account(serverKey: String, userId: String): ProfileDescriptor {
            if (serverKey.length != 64 || serverKey.any { !it.isDigit() && it !in 'A'..'F' } || userId == "00000000-0000-0000-0000-000000000000") {
                throw IllegalArgumentException("Недопустимый идентификатор профиля.")
            }
            val uuid = ru.bgtu_voenmeh.zapara.data.accounts.AccountValidation.id(userId)
            return ProfileDescriptor(false, uuid, serverKey, "profiles/$serverKey/$uuid/zapara.db")
        }
    }
}

File: src/Vograph.Desktop/Services/Profiles/ProfileDescriptor.cs
namespace Vograph.Desktop.Services.Profiles;

public sealed record ProfileDescriptor
{
    private ProfileDescriptor(string root, string? serverKey, Guid? userId)
    {
        GlobalDataDir = Path.GetFullPath(root);
        ServerKey = serverKey;
        UserId = userId;
        DatabasePath = userId is null ? Path.Combine(GlobalDataDir, "vograph.db")
            : Path.Combine(GlobalDataDir, "profiles", serverKey!, userId.Value.ToString("D"), "vograph.db");
    }
    public string GlobalDataDir { get; }
    public string DatabasePath { get; }
    public string? ServerKey { get; }
    public Guid? UserId { get; }
    public bool IsGuest => UserId is null;
    public static ProfileDescriptor Guest(string root) => new(root, null, null);
    public static ProfileDescriptor Account(string root, string key, Guid userId)
    {
        if (key.Length != 64 || key.Any(c => !char.IsAsciiHexDigit(c)) || key != key.ToUpperInvariant() || userId == Guid.Empty)
            throw new ArgumentException("Недопустимый идентификатор профиля.");
        return new(root, key, userId);
    }
}
